frameworks/keyed/hydro-js/src/app.ts

```diff
@@ -0,0 +1,179 @@
+
+const ADJECTIVES = [
+  "pretty",
+  "large",
+  "big",
+  "small",
+  "tall",
+  "short",
+  "long",
+  "handsome",
+  "plain",
+  "quaint",
+  "clean",
+  "elegant",
+  "easy",
+  "angry",
+  "crazy",
+  "helpful",
+  "mushy",
+  "odd",
+  "unsightly",
+  "adorable",
+  "important",
+  "inexpensive",
+  "cheap",
+  "expensive",
+  "fancy",
+];
+const COLOURS = [
+  "red",
+  "yellow",
+  "blue",
+  "green",
+  "pink",
+  "brown",
+  "purple",
+  "brown",
+  "white",
+  "black",
+  "orange",
+];
+const NOUNS = [
+  "table",
+  "chair",
+  "house",
+  "bbq",
+  "desk",
+  "car",
+  "pony",
+  "cookie",
+  "sandwich",
+  "burger",
+  "pizza",
+  "mouse",
+  "keyboard",
+];
+
+const len_ADJECTIVES = ADJECTIVES.length;
+const len_COLOURS = COLOURS.length;
+const len_NOUNS = NOUNS.length;
+
+let nextId = 1;
+
+function buildData(count: number) {
+  const data = new Array(count);
+
+  for (let i = 0; i < count; i++) {
+    data[i] = {
+      id: nextId++,
+      label:
+        ADJECTIVES[random(len_ADJECTIVES)] +
+        " " +
+        COLOURS[random(len_COLOURS)] +
+        " " +
+        NOUNS[random(len_NOUNS)],
+    };
+  }
+
+  return data;
+}
+
+function random(max: number) {
+  return (Math.random() * max) | 0;
+}
+
+setReactivity($("#main")!, { run, runLots, add, update, clear, swapRows });
+
+const data = reactive<Array<{ id: number; label: string }>>([]);
+const selected = reactive(-1);
+
+observe(data, (newData: typeof data, oldData: typeof data) => {
+  merge(newData, oldData);
+  if (!newData.length || !oldData.length) selected(-1);
+});
+
+function add() {
+  data((prev: typeof data) => [...prev, ...buildData(1000)]);
+}
+function run() {
+  data(buildData(1000));
+}
+function runLots() {
+  data(buildData(10000));
+}
+function clear() {
+  data([]);
+}
+function update() {
+  getValue(data).forEach((item: typeof data[number], index: number) => {
+    if (index % 10 === 0) item.label += " !!!";
+  });
+}
+function swapRows() {
+  data((prev: typeof data) => {
+    if (prev.length > 998) {
+      fixSelectedOnSwap(prev);
+
+      [prev[1], prev[998]] = [prev[998], prev[1]];
+    }
+  });
+}
+
+function remove(id: number) {
+  const index = getValue(data).findIndex(
+    (i: typeof data[number]) => i.id === id
+  );
+  data[index].setter(null);
+  data((curr: typeof data) => {
+    curr.splice(index, 1);
+  });
+}
+
+function fixSelectedOnSwap(prev: typeof data) {
+  const second = prev[1].id;
+  const secondLast = prev[998].id;
+  if ([second, secondLast].includes(getValue(selected))) {
+    selected((prev: typeof selected) =>
+      prev === second ? secondLast : second
+    );
+  }
+}
+
+function merge(newData: typeof data, oldData: typeof data) {
+  for (let i = 0; i < oldData.length && newData.length; i++) {
+    oldData[i].id = newData[i].id;
+    oldData[i].label = newData[i].label;
+    newData[i] = oldData[i];
+  }
+
+  for (let i = oldData.length; i < newData.length; i++) {
+    renderItem(newData[i], i);
+  }
+}
```

Comment:
I suspect this implementation is non-keyed. Can't know for certain without running it but it looks like replace at index and render any additional rows.

Reply:
Thanks for having a look. I think it is keyed. A change in the data will automatically change the table. You will see this even better on the clear function. However I do not have a functionality that would add new DOM elements if an item is being inserted. That's why I have to observe the data and add new elements to the table. Would that make it non-keyed?

L152 - L156 could be deleted, they are used as optimization. I am not sure if this is allowed: so if we have 1000 elements already and click on add 1000 elements, it will reuse the existing DOM elements and override the data without destroying the key connection.

Reply:
I see, the `merge` function raised immediate flags with me as it is basically how you would write a non-keyed(index based) reconciler so I thought that was how it was updating. 152-156 is fine assuming it is the same data (same key). Like if I swap 2 rows, with keyed I want to see 2 DOM elements moved. Those TRs literally change locations in the table, not just have the text nodes values change (which is what this assignment looked like).

Other than that without a templating language it is easy to be classified as #772. On one hand you only use DOM commands on creation. The "template" code more or less looks like the compiled output of a reactive friendly view library. On the other hand, the mapping to the DOM nodes is in end user land (ie there is no list helper) which means from an implementation standpoint it is DOM manipulation almost by definition. We've classified this as #772 in the past.

Reply:
Thanks again. I will use a templating language to fix #772. 

I do not have the logic to move elements. If there is a swap, the nodes would change their values, just as said. If this is a requirement for a keyed implementation, then I would move it to non-keyed for now and come up with a solution in the future. Otherwise I probably have to move the elements manually..

Reply:
Yeah I believe this makes it non-keyed. Here is an old article talking about it in more detail: https://www.stefankrause.net/wp/?p=342#:~:text=In%20%E2%80%9Ckeyed%E2%80%9D%20mode%20the%20row,the%20last%20row%20is%20deleted.

Reply:
I moved it to non-keyed and marked it as 772. I would have liked to use the templating language, but for some reason appending the node then will increase the time from 0.22ms to 6ms. The same node, no changes.. Could't figure out why 😔